### PR TITLE
Bluetooth: controller: Implement ADV re-schedule for new scheduler

### PR DESCRIPTION
**Pull request by mtpr-ot**
The new JIT scheduler does not have slot reservation, which means that
the ticker extension feature for automatically re-scheduling a colliding
non-anchored event, e.g. ADV, cannot be used.

This implementaion reacts to ADV envent done with result ABORTED or
TOO_LATE, and in those cases attempts to re-schedule the ADV event again
within the 10 ms pertubation window.

As the original scheduling, the re-scheduling is randomized, so there
is no absolute predictability as to how many attempts will be made. The
advertiser will attempt with randomly delayed re-schdules until the
window is exhausted.

If re-scheduling is unsuccessful, the weight of the ADV event is
increased, improving it's chances of success in the next event.

### CODE REVIEW

File: subsys/bluetooth/controller/ll_sw/lll.h [hunk 1]
```diff
@@ -326,11 +332,12 @@ enum {
 	EVENT_DONE_EXTRA_TYPE_CONN,
 #endif /* CONFIG_BT_CONN */
 
-#if defined(CONFIG_BT_CTLR_ADV_EXT)
+#if defined(CONFIG_BT_CTLR_ADV_EXT) || \
+	defined(CONFIG_BT_CTLR_JIT_SCHEDULING)
```
Comment (wopu-ot) [inline subsys/bluetooth/controller/ll_sw/lll.h:336]: Now that lines can be up to 100 characters, I think this should go onto a single line (also other instances)

**Review by wopu-ot — CHANGES_REQUESTED** [review]
LGTM, just a little formatting issue

File: subsys/bluetooth/controller/ll_sw/lll.h [hunk 2]
```diff
@@ -134,6 +134,12 @@ enum {
 
 #define TICKER_ID_ULL_BASE ((TICKER_ID_LLL_PREEMPT) + 1)
 
+enum done_result {
+	DONE_COMPLETED,
+	DONE_ABORTED,
+	DONE_TOO_LATE
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/lll.h:140]:
```suggestion
	DONE_LATE
```

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c [hunk 3]
```diff
@@ -345,7 +345,18 @@ int lll_done(void *param)
...
+
+	/* TODO: Pass from calling function */
+	result = DONE_COMPLETED;
+
+	lll_done_score(param, result);
+
+	extra = ull_event_done_extra_get();
+	LL_ASSERT(extra);
+
+	extra->result = result;
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c:359]: Add a note, why `extra->type`  is unassigned?

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c [hunk 4]
```diff
@@ -1158,14 +1160,10 @@ static void isr_done(void *param)
...
-	struct event_done_extra *extra;
-
-	extra = ull_event_done_extra_get();
-	LL_ASSERT(extra);
-
-	extra->type = EVENT_DONE_EXTRA_TYPE_ADV;
-#endif  /* CONFIG_BT_CTLR_ADV_EXT */
+#if defined(CONFIG_BT_CTLR_ADV_EXT) || \
+	defined(CONFIG_BT_CTLR_JIT_SCHEDULING)
+	ull_set_done_extra_type(EVENT_DONE_EXTRA_TYPE_ADV);
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c:1165]:
```suggestion
	ull_event_done_extra_type_set(EVENT_DONE_EXTRA_TYPE_ADV);
```

Side note, may be in a separate PR, all other places we can use this function.

File: subsys/bluetooth/controller/ll_sw/ull.c [hunk 5]
```diff
@@ -2509,3 +2512,13 @@ static void disabled_cb(void *param)
 {
 	k_sem_give(param);
 }
+
+void ull_set_done_extra_type(uint8_t type)
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull.c:2516]:
```suggestion
struct event_done_extra *ull_event_done_extra_type_set(uint8_t type)
```

Returning the `struct event_done_extra *` will be useful when other places can reuse this function and then update other fields in the done_extra as necessary.

**Review by cvinayak — CHANGES_REQUESTED** [review]
Overall, no objection to the JIT design for advertising state.

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c (hunk 3 above)
Comment (mtpr-ot) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c:359]: Type is set by the role prior to lll_done. Adding comment to code.

**Review by cvinayak — APPROVED** [review]

**Review by wopu-ot — APPROVED** [review]
